tests/cpp/test_multidevice_transformer.cpp

```diff
@@ -175,6 +175,100 @@ std::vector<at::Tensor> reference_mlp_backwards(
   return grads;
 }
 
+std::vector<at::Tensor> reference_mha_backwards(
+    at::Tensor grad,
+    at::Tensor x,
+    at::Tensor mask,
+    at::Tensor w0,
+    at::Tensor b0,
+    at::Tensor w1,
+    at::ScalarType at_dtype) {
+  // recompute up to sdpa
+  auto m = at::matmul(x, w0).add(b0).view({B, S, 3 * E});
+  auto qkv_vec = m.split(E, /*dim=*/-1);
+  for (auto i = 0; i < 3; i++) {
+    qkv_vec[i] =
+        qkv_vec[i].reshape({B, S, H, E / H}).transpose(1, 2).to(at_dtype);
+  }
+  auto q = qkv_vec[0];
+  auto k = qkv_vec[1];
+  auto v = qkv_vec[2];
+  auto
+      [sdpa_output,
+       log_sumexp,
+       cum_seq_q,
+       cum_seq_k,
+       query_seq_len,
+       key_seq_len,
+       philox_seed,
+       philox_offset,
+       debug_attn_mask] =
+          at::_scaled_dot_product_flash_attention(
+              q,
+              k,
+              v,
+              /*dropout_p=*/kSdpaProb,
+              /*is_causal=*/true,
+              /*return_debug_mask=*/false,
+              /*scale=*/kSdpaScale);
+
+  // backwards pass
+  auto dropout_grad =
+      at::native_dropout_backward(grad, mask, 1.0 / (1.0 - kDropoutProb));
+  auto dropout_grad_q = dropout_grad.to(at_dtype);
+  auto matmul1_grad_x = at::matmul(dropout_grad_q, w1.transpose(0, 1));
+  // sdpa output: B, H, S, E/H
+  auto sdpa_output_reshape = sdpa_output.transpose(1, 2).view({B * S, E});
+  auto matmul1_grad_w =
+      at::matmul(dropout_grad_q.transpose(0, 1), sdpa_output_reshape)
+          .transpose(0, 1);
+  auto matmul1_grad_b = at::sum(dropout_grad, {0});
```

Comment:
```suggestion
  auto linear_grad_b = at::sum(dropout_grad, {0});
```
I slightly prefer calling it linear because matmul, strictly speaking, doesn't take a bias.

Reply:
updated to linear1 which is more consistent with the nvFuser program. Thanks for pointing out the inconsistency! The naming is a bit all over the place... I'll clean up the other functions in a different PR since they also have this issue 😅